### PR TITLE
Create command line arguments to set genesis staking keys, VRF keys, and pool Id

### PR DESCRIPTION
**Pull request by alfiedotwtf**
This is the last item for #1004

### CODE REVIEW

File: common/src/chain/config/mod.rs [hunk 1]
```diff
@@ -567,30 +568,51 @@ fn create_testnet_genesis() -> Genesis {
...
+    let genesis_pool_id = Regex::new("genesis_pool_id:([A-Fa-f0-9]+)")
+        .expect("Regex is ok")
+        .captures(&genesis_staking_settings)
+        .and_then(|c| c.get(1))
+        .map(|c| c.as_str())
+        .or(Some(
+            "123c4c600097c513e088b9be62069f0c74c7671c523c8e3469a1c3f14b7ea2c4",
+        ))
+        .map(decode_hex::<PoolId>)
+        .expect("Parameter decoded {}");
```
Comment (iljakuklic) [inline common/src/chain/config/mod.rs:592]: I find it a bit overkill to use a regex for this. How about just splitting the string on whitespace/comma and using `strip_prefix` to parse key-value entries, roughly like:
```rust
if let Some(pool_id_str) = chunk.strip_prefix("genesis_pool_id:") {
    pool_id = decode_hex(pool_id_str)?;
}
// Same for other settings
```

File: node-lib/src/regtest_options.rs [hunk 2]
```diff
@@ -61,4 +61,8 @@ pub struct ChainConfigOptions {
     /// PoS NetUpgrade override after Genesis
     #[clap(long)]
     pub chain_pos_netupgrades: Option<bool>,
+
+    /// PoS Genesis staking settings
+    #[clap(long)]
+    pub chain_genesis_staking_settings: Option<String>,
```
Comment (iljakuklic) [inline node-lib/src/regtest_options.rs:67]: I believe that if you create a type for the settings, like `GenesisSTakingSettings` and use it here, it will work as long as you provide `FromStr`. May be wrong though.

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py [hunk 3]
```diff
@@ -0,0 +1,428 @@
...
+                        "PublicKey": new_stake_public_key,
+                    },
+                    new_pool_id,
+                ],
+            }
+
+if __name__ == '__main__':
+    # Use the first instance to bootstrap new staking keys
+    GenerateGenesisKeys().main(exit_on_success=False)
+    GeneratePoSBlocksTest().main()
```
Comment (iljakuklic) [inline test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py:428]: Could you please explain what's going on here? Looks like a hack :laughing:

File: common/src/chain/config/mod.rs [hunk 4]
```diff
@@ -567,30 +568,51 @@ fn create_testnet_genesis() -> Genesis {
     )
 }
 
-pub fn regtest_genesis_values() -> (
+pub fn regtest_genesis_values(
+    genesis_staking_settings: Option<String>,
+) -> (
```
Comment (iljakuklic) [inline common/src/chain/config/mod.rs:573]: Would be nicer for this to take some sort of setting type with `Default` impl to avoid the weird special casing for `None`

**Review by iljakuklic — COMMENTED** [review]
A couple of initial observations:

File: common/src/chain/config/mod.rs (hunk 1 above)
Comment (alfiedotwtf) [reply common/src/chain/config/mod.rs]: I'm a regex fan, and it's only called once for the whole process, but no problem.

File: node-lib/src/regtest_options.rs (hunk 2 above)
Comment (alfiedotwtf) [reply node-lib/src/regtest_options.rs]: Nice... I didn't know you could create your own types. Done

File: common/src/chain/config/mod.rs (hunk 4 above)
Comment (alfiedotwtf) [reply common/src/chain/config/mod.rs]: Done

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py (hunk 3 above)
Comment (alfiedotwtf) [reply test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py]: Hack, magic... it's all perspective 😂

File: common/src/chain/config/mod.rs [hunk 5]
```diff
@@ -587,30 +588,74 @@ fn create_testnet_genesis() -> Genesis {
     )
 }
 
-pub fn regtest_genesis_values() -> (
+#[derive(Clone, Debug)]
+pub struct GenesisStakingSettings(String);
```
Comment (iljakuklic) [inline common/src/chain/config/mod.rs:592]: What I meant is something that actually stores the settings in a struct, not just a mere wrapper over a string. The `FromStr` parses the settings string into the struct, including all the error checking and such. The command line parser will then be able to report format errors directly.

File: node-lib/src/regtest_options.rs [hunk 6]
```diff
@@ -61,4 +62,8 @@ pub struct ChainConfigOptions {
     /// PoS NetUpgrade override after Genesis
     #[clap(long)]
     pub chain_pos_netupgrades: Option<bool>,
+
+    /// PoS Genesis staking settings
+    #[clap(long)]
+    pub chain_genesis_staking_settings: Option<GenesisStakingSettings>,
```
Comment (iljakuklic) [inline node-lib/src/regtest_options.rs:68]: Not quite sure about it but you may be able to set something like `default_value = ""` in the `clap` attribute and get rid of the `Option` wrapper.

File: common/src/chain/config/mod.rs [hunk 7]
```diff
@@ -587,30 +588,74 @@ fn create_testnet_genesis() -> Genesis {
...
+    }
+}
+
+impl FromStr for GenesisStakingSettings {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self::new(s.to_string()))
+    }
+}
```
Comment (iljakuklic) [inline common/src/chain/config/mod.rs:616]: My other suggestion notwithstanding, when you have a method that cannot fail, the `std::convert::Infallible` (or equivalent) type should be used as the error type. If errors are possible, try to find or introduce an `enum` that reflects the possible error space.

File: common/src/chain/config/mod.rs [hunk 8]
```diff
@@ -587,30 +588,74 @@ fn create_testnet_genesis() -> Genesis {
...
 }
 
-pub fn regtest_genesis_values() -> (
+#[derive(Clone, Debug)]
+pub struct GenesisStakingSettings(String);
+
+impl GenesisStakingSettings {
+    pub fn new(settings: String) -> Self {
+        Self(settings)
+    }
```
Comment (iljakuklic) [inline common/src/chain/config/mod.rs:597]: If you parse the string in the constructor, you could take a `&str` rather than `String` to initialize the struct.

**Review by iljakuklic — COMMENTED** [review]
Some more comments about input parsing below

**Review by azarovh — COMMENTED** [review]
Agree with Lukas that it makes sense to make `GenesisStakingSettings` a proper struct with fields and parsing and decoding implemented in the constructor.

File: common/src/chain/config/mod.rs (hunk 5 above)
Comment (alfiedotwtf) [reply common/src/chain/config/mod.rs]: Ah, ok. This is done, but I've forgotten to handle the different errors... I'll update this now.

File: node-lib/src/regtest_options.rs (hunk 6 above)
Comment (alfiedotwtf) [reply node-lib/src/regtest_options.rs]: Yep, updated with `default_value_t`.

File: common/src/chain/config/mod.rs [hunk 9]
```diff
@@ -587,30 +588,74 @@ fn create_testnet_genesis() -> Genesis {
...
+
+    pub fn get_settings(&self) -> Vec<&str> {
+        self.0.split(',').collect()
+    }
```
[the remark this answers is not in the document]
Comment (alfiedotwtf) [reply common/src/chain/config/mod.rs]: No longer needed.

File: common/src/chain/config/mod.rs (hunk 7 above)
Comment (alfiedotwtf) [reply common/src/chain/config/mod.rs]: As above, I updated to Infallible, but will change to an enum.

File: common/src/chain/config/mod.rs (hunk 8 above)
Comment (alfiedotwtf) [reply common/src/chain/config/mod.rs]: I keep forgetting this. Fixed.

File: node-lib/src/runner.rs [hunk 10]
```diff
@@ -370,9 +371,14 @@ fn regtest_chain_config(options: &ChainConfigOptions) -> Result<ChainConfig> {
...
 
     if chain_pos_netupgrades.unwrap_or(false) {
-        builder = builder
-            .net_upgrades(NetUpgrades::regtest_with_pos())
-            .genesis_custom(create_regtest_pos_genesis(Destination::AnyoneCanSpend));
+        builder = builder.net_upgrades(NetUpgrades::regtest_with_pos()).genesis_custom(
+            create_regtest_pos_genesis(
+                chain_genesis_staking_settings
+                    .clone()
+                    .unwrap_or(GenesisStakingSettings::default()),
```
[the remark this answers is not in the document]
Comment (alfiedotwtf) [reply node-lib/src/runner.rs]: Yeah, I should have used that... but it's no longer needed now that it's no longer `Option`.

File: common/src/chain/config/regtest.rs [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+            .expect("Private key decoded");
+
+        let vrf_private_key = settings_parts
+            .iter()
+            .find_map(|s| s.strip_prefix("genesis_vrf_private_key:"))
+            .or(Some(
+                "003fcf7b813bec2a293f574b842988895278b396dd72471de2583b242097a59f06e9f3cd7b78d45750afd17292031373fddb5e7a8090db51221038f5e05f29998e",
+            ))
+            .map(decode_hex::<VRFPrivateKey>)
+            .expect("VRF private key decoded");
```
Comment (iljakuklic) [inline common/src/chain/config/regtest.rs:73]: I am a bit conflicted about the `.expect`s here. I know this is supposed to be used for testing only. However, it is still being parsed from user input from the command line, so some sort of nicer error reporting is probably preferable.

A related question is what to do about unrecognised keys. Let's say the user makes a typo `genesis_vrf_pirvate_key:BLAH`, then vrf key is silently going to be set to the default value instead of `BLAH`. That may be quite annoying to diagnose.

File: common/src/chain/config/regtest.rs [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    vrf_private_key: VRFPrivateKey,
+}
+
+impl GenesisStakingSettings {
+    pub fn new(settings: &str) -> Self {
+        let settings_parts: Vec<&str> = settings.split(',').map(|s| s.trim()).collect();
+
+        let pool_id = settings_parts
+            .iter()
+            .find_map(|s| s.strip_prefix("genesis_pool_id:"))
```
Comment (iljakuklic) [inline common/src/chain/config/regtest.rs:50]: Probably a matter of taste to some degree, but I find the `genesis_` prefix in the option names quite redundant. The type already has `Genesis` in the name. When taken from the command line, the option is called `--chain-genesis-staking-settings`. It seems to me dropping the word "genesis" here would not cause any ambiguities assuming you don't plan to generalise this in the future to include non-genesis stuff too.

File: common/src/chain/config/regtest.rs (hunk 11 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs:73]: > I am a bit conflicted about the .expects here

Yep agreed. I've already removed these locally.

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+                                "index": 1,
+                            },
+                        },
+                ],
+                 "kernel_input_utxo": [
+                    {
+                        "CreateStakePool": [
+                            genesis_pool_id,
+                            {
+                                "value": "4000000000000000",
```
Comment (TheQuantumPhysicist) [inline test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py:234]: Perhaps we should find a proper way to calculate coins from atoms instead this chaotic zeros' situation.

File: common/src/chain/config/regtest.rs [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+            .or(Some(
+                "123c4c600097c513e088b9be62069f0c74c7671c523c8e3469a1c3f14b7ea2c4",
```
Comment (TheQuantumPhysicist) [inline common/src/chain/config/regtest.rs:52]: If these are default values, please move them to constants.

File: test-rpc-functions/src/rpc.rs [hunk 15]
```diff
@@ -97,7 +99,8 @@ trait RpcTestFunctionsRpc {
 #[async_trait::async_trait]
 impl RpcTestFunctionsRpcServer for super::RpcTestFunctionsHandle {
     async fn genesis_pool_id(&self) -> rpc::Result<Option<String>> {
-        let (genesis_pool_id, genesis_stake_pool_data, _, _, _, _) = regtest_genesis_values();
+        let (genesis_pool_id, genesis_stake_pool_data, _, _, _, _) =
+            genesis_values(GenesisStakingSettings::new(""));
```
Comment (TheQuantumPhysicist) [inline test-rpc-functions/src/rpc.rs:103]: If this is the default value, it's preferable to use the Default::default() trait function to indicate what this really is.

File: node-lib/src/regtest_options.rs [hunk 16]
```diff
@@ -61,4 +62,8 @@ pub struct ChainConfigOptions {
     /// PoS NetUpgrade override after Genesis
     #[clap(long)]
     pub chain_pos_netupgrades: Option<bool>,
+
+    /// PoS Genesis staking settings
+    #[clap(long, default_value_t = GenesisStakingSettings::new(""))]
```
Comment (TheQuantumPhysicist) [inline node-lib/src/regtest_options.rs:67]: default()

File: common/src/chain/config/regtest.rs [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+
+    let premine_output =
+        TxOutput::Transfer(OutputValue::Coin(DEFAULT_INITIAL_MINT), premine_destination);
+
+    Genesis::new(
+        String::new(),
+        BlockTimestamp::from_int_seconds(1639975460),
+        vec![premine_output, create_genesis_pool_txoutput],
+    )
+}
```
Comment (TheQuantumPhysicist) [inline common/src/chain/config/regtest.rs:180]: Some basic unit tests that explain the parsing would really help.

You don't have to test every corner case (please don't spend days creating them), but usually unit tests can be used by beginners to learn how a struct works.

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py [hunk 18]
```diff
@@ -0,0 +1,438 @@
...
+                    },
+                    new_pool_id,
+                ],
+            }
+
+if __name__ == '__main__':
+    # We need to spin up a node in order to use the "test-functions/"
+    # RPC endpoint so we can generate new genesis staking
+    # settings. However, to spin up a node, the framework will create
+    # a chaintate along with its own Genesis, defeating the purpose of
```
Comment (TheQuantumPhysicist) [inline test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py:429]: chaintate

**Comment by TheQuantumPhysicist** [comment]
This is looking fine. I'll review again after addressing the comments.

File: common/src/chain/config/regtest.rs (hunk 11 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs]: This was a tricky one...

`Display` is required for `Clap` for custom types but it doesn't really elaborate why. I thought it was just using this for user `--help` etc so I just did a `format!("{:?}...)`, but then everything broke. It turned out that when not provided on the command line, it uses `Display` to stringify `default()`, and then calls from_str() later on to parse the value.

So when passed on the command line, `GenesisStakingSettings::from_str(settings: &str)` would have `settings` set to whatever was taken from the command line, however when not passed on the command line, `settings` contained something like `{pool_id: Id<Pool>{0x123c4c600097c513e088b9be62069f0c74c7671c523c8e3469a1c3f14b7ea2c4}, stake_private_key: PrivateKey { key: Secp256k1Schnorr(Secp256k1PrivateKey { data: SecretKey(#0fd5243c90a0082f) }) },...`, all because my `format!()` in the `impl Display` wasn't "round tripping" the value as stated in the `Clap` documentation.

All good, it's working now.

File: common/src/chain/config/regtest.rs (hunk 12 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs]: Done.

File: common/src/chain/config/regtest.rs (hunk 14 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs]: Cool, done.

File: test-rpc-functions/src/rpc.rs (hunk 15 above)
Comment (alfiedotwtf) [reply test-rpc-functions/src/rpc.rs]: Fixed.

File: node-lib/src/regtest_options.rs (hunk 16 above)
Comment (alfiedotwtf) [reply node-lib/src/regtest_options.rs]: Turns out `default_value_t` doesn't need an actual parameter when there's a `default()`.

File: common/src/chain/config/regtest.rs (hunk 17 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs:234]: No problem. Added.

`mod.rs`s' `decode_hex()` was breaking these tests because it panicks when decoding fails. Because I didn't want to change the upstream function, I've copied it into `regtest.rs` and made that return `Result`. If you want me to change upstreams to return a `Result` too and remove the duplicated code, let me know.

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py (hunk 18 above)
Comment (alfiedotwtf) [reply test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py]: Fixed.

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py (hunk 13 above)
Comment (alfiedotwtf) [reply test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py:234]: I've added it as an item to #1013

File: common/src/chain/config/regtest.rs [hunk 19]
```diff
@@ -0,0 +1,430 @@
...
+
+impl FromStr for GenesisStakingSettings {
+    type Err = GenesisStakingSettingsInputErrors;
+
+    fn from_str(settings: &str) -> Result<Self, Self::Err> {
+        let mut settings_parts = settings
+            .split(',')
+            .filter_map(|s| s.split_once(':'))
+            .map(|(k, v)| (k.trim(), v.trim()))
+            .collect::<HashMap<&str, &str>>();
```
Comment (iljakuklic) [inline common/src/chain/config/regtest.rs:116]: The `filter_map` call will silently discard entries without a comma. Would it be better to report an error instead?

Also collecting into HashMap will handle duplicate keys in the "last one wins" fashion. Probably not a big deal but maybe consider making that into an error too?

File: common/src/chain/config/regtest.rs [hunk 20]
```diff
@@ -0,0 +1,430 @@
...
+    key::{PrivateKey, PublicKey},
+    vrf::{VRFPrivateKey, VRFPublicKey},
+};
+use hex::FromHex;
+use serialization::hex::HexEncode;
+
+const GENESIS_POOL_ID: &str = "123c4c600097c513e088b9be62069f0c74c7671c523c8e3469a1c3f14b7ea2c4";
+const GENESIS_STAKE_PRIVATE_KEY: &str =
+    "008717e6946febd3a33ccdc3f3a27629ec80c33461c33a0fc56b4836fcedd26638";
+const GENESIS_VRF_PRIVATE_KEY: &str = "003fcf7b813bec2a293f574b842988895278b396dd72471de2583b242097a59f06e9f3cd7b78d45750afd17292031373fddb5e7a8090db51221038f5e05f29998e";
```
Comment (iljakuklic) [inline common/src/chain/config/regtest.rs:42]: Could these be proper types, like `PoolId`, `PrivateKey`, `VRFPrivateKey` instead of strings? You can use the `hex!` macro from the `hex` crate to enter hex strings and get a `[u8; N]`.

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py [hunk 21]
```diff
@@ -0,0 +1,438 @@
...
+                            },
+                            "index": 0,
+                        },
+                    }
+                ],
+                "outputs": [
+                    {
+                        "Transfer": [
+                            {
+                                "Coin": "10000000000000000",
```
Comment (iljakuklic) [inline test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py:269]: Nitpick: I think you can enter a number here instead of a string, and use the underscore syntax for better readability. Like `10_000_000_000_000_000`. Not sure it will fit into Python's numeric type but I suspect it uses some sort of bignum by default?

**Review by iljakuklic — APPROVED** [review]
Looks good. A couple more minor suggestions here:

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py (hunk 21 above)
Comment (TheQuantumPhysicist) [reply test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py:269]: While I don't disagree there, the problem is that the number of decimals isn't a multiple of three, so this won't help to make the number more understandable. We should use proper COIN factor, being 10^11 or whatever correct decimal count, then we have 10_000*COIN.

File: node-lib/src/runner.rs [hunk 22]
```diff
@@ -333,6 +333,7 @@ fn regtest_chain_config(options: &ChainConfigOptions) -> Result<ChainConfig> {
         chain_max_block_size_with_standard_txs,
         chain_max_block_size_with_smart_contracts,
         chain_pos_netupgrades,
+        chain_genesis_staking_settings,
```
Comment (TheQuantumPhysicist) [inline node-lib/src/runner.rs:336]: We should add guards such that if someone tries to pass staking settings, net upgrades, emission schedule, coin decimals, target spacing, and any other feature that's made for testing, in anything other than regtest, then we crash the software. This is important to prevent scammers from trying to get victims to create fake nodes somehow with custom settings and make them feel they got rich or something.

File: common/src/chain/config/regtest.rs (hunk 17 above)
Comment (TheQuantumPhysicist) [reply common/src/chain/config/regtest.rs:234]: Are you talking about [#1053](https://github.com/mintlayer/mintlayer-core/issues/1053)?

**Review by TheQuantumPhysicist — APPROVED** [review]
Besides the comments, I think this is looking shiny enough.

File: common/src/chain/config/regtest.rs (hunk 17 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs:234]: No, this was different. For example, in the `invalid_vrf_private_key()` test, because the input wasn't long enough nor hex, `mod.rs`s `decode_hex()` would panic and so my test wouldn't pass.

File: common/src/chain/config/regtest.rs (hunk 20 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs:42]: I think it's best to keep them as hex strings, because this is the same format as what is set on the command line, and also to be consistent with the other networks (defined in `mod.rs`).

File: node-lib/src/runner.rs (hunk 22 above)
Comment (alfiedotwtf) [reply node-lib/src/runner.rs:336]: It looks like `node-lib/src/regtest_options.rs`, where all of these "dangerous" parameters are defined to be read from config, are only used in RegTest (from `node-lib/src/runner.rs`). They can also be set from code (i.e not the command line) using the Builder, but to play it safe I'll create an issue to check the network type when creating a new `ChainConfigOptions` instances

File: common/src/chain/config/regtest.rs (hunk 19 above)
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs:116]: The `split(',')` will return an array with the empty string, so will proceed through the chain, but you're right about the `filter_map()` on settings lacking `':'`. This is now fixed.

File: test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py (hunk 21 above)
Comment (alfiedotwtf) [reply test/functional/blockprod_generate_pos_blocks_rand_genesis_keys.py:269]: Using `COIN` a 10^11, this is now fixed.

File: common/src/chain/config/regtest.rs [hunk 23]
```diff
@@ -0,0 +1,430 @@
...
+            vrf_private_key,
+        }
+    }
+
+    pub fn pool_id(&self) -> PoolId {
+        self.pool_id
+    }
+
+    pub fn stake_private_key(&self) -> PrivateKey {
+        self.stake_private_key.clone()
```
[the remark this answers is not in the document]
Comment (alfiedotwtf) [reply common/src/chain/config/regtest.rs:94]: Thanks, I always forget to do this and opt for `clone()`. Fixed.

File: Cargo.toml [hunk 24]
```diff
@@ -159,6 +159,7 @@ quote = "1.0"
 rand = "0.8"
 rand_chacha = "0.3"
 reedline = "0.22"
+regex = "1.9"
```
Comment (iljakuklic) [inline Cargo.toml:162]: Is this still used?

**Review by iljakuklic — APPROVED** [review]

File: Cargo.toml (hunk 24 above)
Comment (alfiedotwtf) [reply Cargo.toml:162]: I removed it and it broke. Now that I've rebased I'll try it again

Comment (alfiedotwtf) [reply Cargo.toml:162]: Good pickup. It snuck back in because a few rebases failed. I'm surprised Clippy didn't complain. Fixed.

Comment (TheQuantumPhysicist) [reply Cargo.toml]: I have a huge problem with regex in security applications. It's the source of all evil.

File: node-lib/src/runner.rs (hunk 22 above)
Comment (alfiedotwtf) [reply node-lib/src/runner.rs:343]: Because of the way how our command line parsing works, if any of the regtest command line arguments are passed along to other networks, the cli will fail e.g `error: unexpected argument '--chain-pos-netupgrades' found` so I think this should already prevent non-regtest use.

However, https://github.com/mintlayer/mintlayer-core/pull/1092/commits/51789930a744537499582b420bd0994d5436ea03 adds a gate to the regtest options parsing, so at least now in code you must explicitly be passing in the `Command::RegTest`.

File: test/functional/blockprod_generate_pos_blocks.py [hunk 25]
```diff
@@ -25,6 +25,8 @@
 
 import random, time
 
+COIN = 100_000_000_000
```
Comment (TheQuantumPhysicist) [inline test/functional/blockprod_generate_pos_blocks.py:28]: This should be a global definition somewhere. I bet there's already a definition somewhere for bitcoin. We shouldn't redefine it for every test.

File: node-lib/src/runner.rs (hunk 22 above)
Comment (TheQuantumPhysicist) [reply node-lib/src/runner.rs:343]: Looking good!
